ansible_collections/ds389/ansible_ds/tests/unit/plugins/modules/test_state.py

```diff
@@ -0,0 +1,525 @@
+                         'inst2', ],
+        },
+        {
+            'msg': 'after marking a ds389_agmt as absent',
+            'avas': [('fulltargetname', 'localhost.inst5.be502')],
+            'removed': [ 'ds389_agmt_be102_localhost.inst5.be502', ],
+        },
+    ]
+
+    def _init_test(self):
+        # Initialize 389ds instances  as if _AbsentHandler tests have run
+        for action in _AbsentHandler.ACTIONS:
+            tgt = self.get_target(self.args, action['avas'])
+            tgt['state'] = 'absent'
+            
+            if tgt:
+                self.remove_target(self.args, tgt)
+            for elmt in action['removed']:
+                self.expected_entities.remove(elmt)
+        tgt = self.get_target(self.args,[('state', 'absent'),])
+        while tgt:
+            self.remove_target(tgt)
+            tgt = self.get_target(self.args,[('state', 'absent'),])
+        self.log.info('Initialize 389ds instance using ds389_server module')
+        result = self.ansibletest.run_test_module( self.args )
+        self.verify_entities(result, "after initialization")
+        # Set state to updated.
+        self.args["ANSIBLE_MODULE_ARGS"]["ds389"]["state"] = "updated"
+
+    def run(self, action):
+        self.log.info(f"Run ds389_server module {action['msg']}")
+        tgt = self.get_target(self.args, action['avas'])
+        for elmt in action['removed']:
+            if elmt in self.expected_entities:
+                self.expected_entities.remove(elmt)
+            else:
+                self.log.debug(f'_UpdatedHandler: Cannot remove expected_entities {elmt} in action {action}')
+                self.ok = False
+        self.remove_target(self.args, tgt)
+        #self.log.debug(f'_UpdatedHandler.run: action={action} removed tgt = {tgt} self.args={self.args}')
+        result = self.ansibletest.run_test_module( self.args )
+        self.verify_entities(result, action['msg'])
+
+
+@pytest.mark.parametrize(
+    "testname",
+    ( "present", "absent", "updated" )
+)
+def test_states(ansibletest, testname):
+    """Test ds389_server module by changing state of various entities.
+        Setup: None
+        Step 1: Creates initial configuration
+        Step 2: Run test actions then verify the entity lists:
+               - add/remove an index
+               - add/remove an agmt
+               - add/remove a backend
+               - add/remove an instance
+               - add/remove a ds389_agmt
+        Result 1: No error
+        Result 2: No error
+    """
+
+    log = ansibletest.get_log(__name__)
+    try:
+        action_handler = _StateTestHandler.get_handler(testname)
+        action_handler.init_test(ansibletest, log)
+        for action in action_handler.ACTIONS:
+            action_handler.run(action)
+        action_handler.end_test()
+        assert action_handler.ok
+    except ( AssertionError, ldap.LDAPError ) as exc:
+        action_handler.ansibletest.save_artefacts()
+        raise exc
```

Comment:
I see souple of possible issues here:
1. `action_handler` may not be there because of some weird error. So I think it makes sense to add check here:
```
if action_handler:
            action_handler.ansibletest.save_artefacts()
```
2. If any other exception will happen, we won't know. So we probably should deal with that too:
```
    except Exception as exc:
```

Reply:
No, it is a plain bug:  action_handler has no save_artefacts method 
    I intended to write  directly ansibletest.save_artefacts() ...